lib/circuits_sim/device/sht4x.ex

```diff
@@ -0,0 +1,128 @@
+  def set_humidity_rh(bus_name, address, value) when is_number(value) do
+    I2CServer.send_message(bus_name, address, {:set_humidity_rh, value})
+  end
+
+  @spec set_temperature_c(String.t(), Circuits.I2C.address(), float()) :: :ok
+  def set_temperature_c(bus_name, address, value) when is_number(value) do
+    I2CServer.send_message(bus_name, address, {:set_temperature_c, value})
+  end
+
+  ## protocol implementation
+
+  defimpl I2CDevice do
+    @impl I2CDevice
+    def read(%{current: :serial_number} = state, count) do
+      result = binary_for_serial_number(state) |> trim_pad(count)
+      {result, %{state | current: nil}}
+    end
+
+    def read(%{current: :measure_high_repeatability} = state, count) do
+      result = raw_sample(state) |> trim_pad(count)
+      {result, %{state | current: nil}}
+    end
+
+    def read(%{current: :measure_medium_repeatability} = state, count) do
+      result = raw_sample(state) |> trim_pad(count)
+      {result, %{state | current: nil}}
+    end
+
+    def read(%{current: :measure_low_repeatability} = state, count) do
+      result = raw_sample(state) |> trim_pad(count)
+      {result, %{state | current: nil}}
+    end
+
+    def read(state, count) do
+      {:binary.copy(<<0>>, count), %{state | current: nil}}
+    end
+
+    @impl I2CDevice
+    def write(state, <<0x89>>), do: %{state | current: :serial_number}
+    def write(state, <<0xFD>>), do: %{state | current: :measure_high_repeatability}
+    def write(state, <<0xF6>>), do: %{state | current: :measure_medium_repeatability}
+    def write(state, <<0xE0>>), do: %{state | current: :measure_low_repeatability}
+    def write(state, _), do: state
+
+    @impl I2CDevice
+    def write_read(state, _to_write, read_count) do
+      {:binary.copy(<<0>>, read_count), %{state | current: nil}}
+    end
+
+    defp trim_pad(x, count) when byte_size(x) >= count, do: :binary.part(x, 0, count)
+    defp trim_pad(x, count), do: x <> :binary.copy(<<0>>, count - byte_size(x))
+
+    @impl I2CDevice
+    def render(state) do
+      humidity_rh = Float.round(state.humidity_rh, 3)
+      temperature_c = Float.round(state.temperature_c, 3)
+      "Temperature: #{temperature_c}°C, Relative humidity: #{humidity_rh}%"
+    end
+
+    @impl I2CDevice
+    def handle_message(state, {:set_humidity_rh, value}) do
+      {:ok, %{state | humidity_rh: value}}
+    end
+
+    def handle_message(state, {:set_temperature_c, value}) do
+      {:ok, %{state | temperature_c: value}}
+    end
+
+    defp binary_for_serial_number(state) do
+      <<state.serial_number::32>> |> add_crcs()
+    end
+
+    defp raw_sample(state) do
+      raw_rh = round((state.humidity_rh + 6) * (0xFFFF - 1) / 125)
+      raw_t = round((state.temperature_c + 45) * (0xFFFF - 1) / 175)
+      <<raw_t::16, raw_rh::16>> |> add_crcs()
+    end
+
+    defp add_crcs(data) do
+      for <<uint16::16 <- data>>, into: <<>> do
+        crc = SHT4X.Calc.checksum(<<uint16::16>>)
```

Comment:
The checksum implementation needs to be copied here since `SHT4X` can't be called. To see this, try running `MIX_ENV=prod iex -S mix`.